java-checks/src/main/java/org/sonar/java/checks/regex/EmptyLineRegexCheck.java

```diff
@@ -0,0 +1,190 @@
+    .addParametersMatcher(JAVA_LANG_STRING, JAVA_LANG_STRING)
+    .build();
+
+  private static final MethodMatchers PATTERN_COMPILE = MethodMatchers.create()
+    .ofTypes(JAVA_UTIL_PATTERN)
+    .names("compile")
+    .addParametersMatcher(JAVA_LANG_STRING)
+    .addParametersMatcher(JAVA_LANG_STRING, "int")
+    .build();
+
+  private static final MethodMatchers PATTERN_MATCHER = MethodMatchers.create()
+    .ofTypes(JAVA_UTIL_PATTERN)
+    .names("matcher")
+    .addParametersMatcher("java.lang.CharSequence")
+    .build();
+
+  private static final MethodMatchers PATTERN_FIND = MethodMatchers.create()
+    .ofTypes("java.util.regex.Matcher")
+    .names("find")
+    .addWithoutParametersMatcher()
+    .build();
+
+  private static final MethodMatchers STRING_IS_EMPTY = MethodMatchers.create()
+    .ofTypes(JAVA_LANG_STRING)
+    .names("isEmpty")
+    .addWithoutParametersMatcher()
+    .build();
+
+  @Override
+  public List<Tree.Kind> nodesToVisit() {
+    return Collections.singletonList(Tree.Kind.METHOD_INVOCATION);
+  }
+
+  @Override
+  public void visitNode(Tree tree) {
+    MethodInvocationTree mit = (MethodInvocationTree) tree;
+    if (PATTERN_COMPILE.matches(mit)) {
+      checkPatternCompile(mit);
+    } else if (STRING_REPLACE.matches(mit)) {
+      checkStringReplace(mit);
+    }
+  }
+
+  private void checkPatternCompile(MethodInvocationTree mit) {
+    ExpressionTree firstArgument = mit.arguments().get(0);
+
+    if (isEmptyLineMultilineRegex(firstArgument)) {
+      reportIfUsedOnEmpty(mit, firstArgument);
+    } else if (mit.arguments().size() == 2) {
+      ExpressionTree secondArgument = mit.arguments().get(1);
+      if (isEmptyLineRegex(firstArgument) && isMultilineFlag(secondArgument)) {
+        reportIfUsedOnEmpty(mit, secondArgument);
+      }
+    }
+  }
+
+  private void checkStringReplace(MethodInvocationTree mit) {
+    ExpressionTree firstArgument = mit.arguments().get(0);
+    ExpressionTree methodSelect = mit.methodSelect();
+    if (isEmptyLineMultilineRegex(firstArgument)
+    && methodSelect.is(Tree.Kind.MEMBER_SELECT)
+    && canBeEmpty(((MemberSelectExpressionTree) methodSelect).expression())) {
+      reportIssue(firstArgument, MESSAGE);
+    }
+  }
+
+  private static boolean isEmptyLineMultilineRegex(ExpressionTree regexArgument) {
+    return regexArgument.asConstant(String.class).filter(EMPTY_LINE_MULTILINE_REGEX::equals).isPresent();
+  }
+
+  private static boolean isEmptyLineRegex(ExpressionTree regexArgument) {
+    return regexArgument.asConstant(String.class).filter(EMPTY_LINE_REGEX::equals).isPresent();
+  }
+
+  private static boolean isMultilineFlag(ExpressionTree flag) {
+    if (flag.is(Tree.Kind.MEMBER_SELECT)) {
+      MemberSelectExpressionTree memberSelect = (MemberSelectExpressionTree) flag;
+      return memberSelect.expression().symbolType().isSubtypeOf(JAVA_UTIL_PATTERN)
+        && "MULTILINE".equals(memberSelect.identifier().name());
+    }
```

Comment:
This won't correctly identify the flag if multiple flags are used (`Pattern.compile(bla, Pattern.MULTILINE | Patter.COMMENT`). As it turns out, this is currently broken anyway, but I'm about to submit a PR that makes AbstractRegexCheck correctly handle ORed flags like that.